### ISSUE TITLE
allow conversion of images to a different format, and allow non-web formats

### ISSUE THREAD

**Issue by xlotlu** [lektor/lektor#607]
Whether creating a thumbnail or not, one should be able to request a specific image format, and  not be forced to have the source image in jpeg/png/gif format in the first place.

Scenario: I am a photographer who keeps his collection in TIFF format. I want to publish this on the web with lektor, but without pre-converting the entire collection to a web-friendly format. I want to be able to use this in my template:

```html
<a href="{{ img.convert(format='jpeg').url }}"><img src="{{ img.thumbnail(width=100, height=100, format='jpeg').url }}"></a>
```

This issue implies:
- allow the source image to be other non-web-friendly formats. Such an image object should be a shim, there should be no need to support dimensions retrieval etc. for them. (at least in the beginning, anyway)
- a new image method that allows only changing format, I'm thinking `convert`. Or maybe `convert` is the full-fledged thing and `thumbnail` could become an alias for it.

Also in the scope of this issue, `.thumbnail()` should debatably default to JPEG mode, as one does not usually want lossless thumbnails.

**Comment by ikus060** [comment]
I would clearly want this too. thumbnail() is currently generating a lossless png that is ~1.3Mb. That is not really a thumbnail. I would like to force thumbnail to a specific image format.

Looking at the code it's seams relatively easy, to change.

Any guidance before I start writing a PR for this ?

**Comment by runfalk** [comment]
It should indeed be fairly straight forward. The biggest question is whether we should switch the default to JPEG or not. I think it doesn't really make sense to match the source format, and JPEG should probably be the default, but other's should probably weigh in on this. This might also be a good time to add webp support (I think that landed in ImageMagick 7).

You should start here https://github.com/lektor/lektor/blob/f273c6918b5f1043a6b09980de0a82f758033929/lektor/db.py#L807 and add an extra kwarg for `format`. Other than that read this https://github.com/lektor/lektor#want-to-develop-on-lektor and make sure you add a unit test.

**Comment by dairiki** [comment]
> The biggest question is whether we should switch the default to JPEG or not. I think it doesn't really make sense to match the source format, and JPEG should probably be the default, but others should probably weigh in on this.

My gut feeling is to leave the default as it is (thumbnail format matches input format).

[Explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/#id2). It's not too onerous to add a `format='jpeg'` to ones `.thumbnail` calls if one wants to change the format.

For the sake of compatibility, changing the default behavior should be avoided when reasonably possible (since it affects existing sites.)

**Comment by xlotlu** [comment]
> For the sake of compatibility, changing the default behavior should be avoided when reasonably possible (since it affects existing sites.)

That's true, but in this particular case it affects existing sites in a positive way. :)

> This might also be a good time to add webp support (I think that landed in ImageMagick 7).

How about a global config item, say "`default_thumbnail_format`", that defaults to `None` in the beginning? `None` means preserve format, while anything explicit, jpeg or webp applies site-wide. That covers all angles: a sane default, user preference, *and* backwards compatibility.

**Comment by runfalk** [comment]
> How about a global config item, say "default_thumbnail_format", that defaults to None in the beginning?

I like that idea. We could put a deprecation warning that tells the user the behavior will change in a future release and that they should set it to `match_source` (or something) if they want to keep current behavior.
